File: admin/src/api/marketing.ts
import request from '@/utils/request'

export interface PageResponse<T> {
  content: T[]
  size: number
  page: number
  total_elements: number
  total_pages: number
}

export interface MarketingLead {
  id: number
  email: string
  name?: string
  source?: string
  status: string
  unsubscribed: boolean
  created_at: string
  last_seen_at?: string
}

export interface MarketingCampaign {
  id: number
  name: string
  subject: string
  landing_url: string
  status: string
  provider_receiver_id?: string
  provider_template_id?: string
  provider_task_id?: string
  created_at: string
  delivery_count: number
}

export interface ImportLeadsResp {
  created: number
  duplicated: number
  invalid: number
}

export interface FunnelMetric {
  event_type: string
  count: number
  rate: number
}

export interface CampaignFunnel {
  campaign_id: number
  sent: number
  metrics: FunnelMetric[]
}

export interface MarketingEvent {
  id: number
  campaign_id?: number
  delivery_id?: number
  lead_id?: number
  event_type: string
  url?: string
  user_agent?: string
  created_at: string
  meta?: Record<string, unknown>
}

export const getMarketingLeads = (params: {
  page?: number
  size?: number
  keyword?: string
  source?: string
  unsubscribed?: boolean
}): Promise<PageResponse<MarketingLead>> => {
  return request.get('/admin/marketing/leads', { params })
}

export const importMarketingLeads = (data: {
  csv_content: string
  source?: string
}): Promise<ImportLeadsResp> => {
  return request.post('/admin/marketing/leads/import', data)
}

export const getMarketingCampaigns = (params: {
  page?: number
  size?: number
  keyword?: string
  status?: string
}): Promise<PageResponse<MarketingCampaign>> => {
  return request.get('/admin/marketing/campaigns', { params })
}

export const createMarketingCampaign = (data: {
  name: string
  subject: string
  landing_url: string
  provider_template_id: string
  lead_ids: number[]
}): Promise<MarketingCampaign> => {
  return request.post('/admin/marketing/campaigns', data)
}

export const sendMarketingCampaign = (id: number): Promise<{
  receiver_id: number
  task_id: number
  delivery_count: number
}> => {
  return request.post(`/admin/marketing/campaigns/${id}/send`)
}

export const getCampaignFunnel = (id: number): Promise<CampaignFunnel> => {
  return request.get(`/admin/marketing/campaigns/${id}/funnel`)
}

export const getCampaignEvents = (id: number): Promise<MarketingEvent[]> => {
  return request.get(`/admin/marketing/campaigns/${id}/events`)
}

File: src/views/marketing.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use validator::Validate;

#[derive(Debug, Deserialize)]
pub struct LeadQuery {
    pub keyword: Option<String>,
    pub source: Option<String>,
    pub unsubscribed: Option<bool>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct ImportLeadsReq {
    #[validate(length(min = 1, max = 5_000_000, message = "CSV 内容不能为空或过大"))]
    pub csv_content: String,
    #[validate(length(max = 80, message = "来源不能超过80个字符"))]
    pub source: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ImportLeadsResp {
    pub created: usize,
    pub duplicated: usize,
    pub invalid: usize,
}

#[derive(Debug, Serialize)]
pub struct MarketingLeadResp {
    pub id: i64,
    pub email: String,
    pub name: Option<String>,
    pub source: Option<String>,
    pub status: String,
    pub unsubscribed: bool,
    pub created_at: String,
    pub last_seen_at: Option<String>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct CreateCampaignReq {
    #[validate(length(min = 1, max = 120, message = "活动名称长度必须在1-120字符之间"))]
    pub name: String,
    #[validate(length(min = 1, max = 200, message = "邮件主题长度必须在1-200字符之间"))]
    pub subject: String,
    #[validate(length(min = 1, max = 500, message = "Landing URL 长度必须在1-500字符之间"))]
    pub landing_url: String,
    #[validate(length(min = 1, max = 100, message = "腾讯云 SES 模板 ID 不能为空"))]
    pub provider_template_id: String,
    #[validate(length(min = 1, max = 10000, message = "请选择1-10000个线索"))]
    pub lead_ids: Vec<i64>,
}

#[derive(Debug, Serialize)]
pub struct MarketingCampaignResp {
    pub id: i64,
    pub name: String,
    pub subject: String,
    pub landing_url: String,
    pub status: String,
    pub provider_receiver_id: Option<String>,
    pub provider_template_id: Option<String>,
    pub provider_task_id: Option<String>,
    pub created_at: String,
    pub delivery_count: u64,
}

#[derive(Debug, Deserialize)]
pub struct CampaignQuery {
    pub keyword: Option<String>,
    pub status: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct SendCampaignResp {
    pub receiver_id: i64,
    pub task_id: i64,
    pub delivery_count: u64,
}

#[derive(Debug, Deserialize)]
pub struct MarketingEventReq {
    pub event_type: String,
    pub mtk: String,
    pub url: Option<String>,
    pub meta: Option<Value>,
}

#[derive(Debug, Serialize)]
pub struct MarketingEventResp {
    pub recorded: bool,
}

#[derive(Debug, Serialize)]
pub struct FunnelMetric {
    pub event_type: String,
    pub count: usize,
    pub rate: f64,
}

#[derive(Debug, Serialize)]
pub struct CampaignFunnelResp {
    pub campaign_id: i64,
    pub sent: usize,
    pub metrics: Vec<FunnelMetric>,
}
